075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+
+## Motivation
+
+The rebalancing operation during scaling up and down is a manual two steps process.
+
+* scaling the cluster up by adding brokers, then running a rebalance.
+* running a rebalance, then scaling the cluster down by removing brokers.
+
+In order to simplify the process, it would be great for the user to deal with just scaling the cluster and then having the operator taking care of running a rebalancing operation automatically when needed.
+
+## Proposal
+
+This proposal is about providing the user the possibility to automatically run a rebalancing when they scale the cluster up (by adding brokers) or scale the cluster down (by removing brokers) by changing the `spec.replicas` within a `Kafka` or `KafkaNodePool` custom resource.
+
+### Extending the `Kafka.spec.cruiseControl` field with auto-rebalancing related configuration
+
+The `Kafka` custom resource is extended by adding a new `spec.cruiseControl.autoRebalance` field.
+By default, without such field, no auto-rebalancing runs on scaling.
+
+The `spec.cruiseControl.autoRebalance` field allows to specify a list of rebalancing modes with the corresponding configuration to be used.
+For each entry in the list, the mode is defined by the `spec.cruiseControl.autoRebalance.mode` field (i.e. `add-brokers` or `remove-brokers`) and the corresponding rebalancing configuration is defined as a reference to a `KafkaRebalanceSettings` custom resource by using the `spec.cruiseControl.autoRebalance.settings` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/local-object-reference/).
+This field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration.
+
+In the following example, a user can decide to use the same rebalancing settings when adding or removing brokers.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using the same rebalancing settings when adding or removing brokers
+      - mode: add-brokers
+        settings: my-add-remove-brokers-rebalancing-settings
+      - mode: remove-brokers
+        settings: my-add-remove-brokers-rebalancing-settings
+```
+
+The user can even decide to use different settings when adding or removing brokers.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using specific settings when adding brokers
+      - mode: add-brokers
+        settings: my-add-brokers-rebalancing-settings
+      # using different settings when removing brokers  
+      - mode: remove-brokers
+        settings: my-remove-brokers-rebalancing-settings
+```
+
+The user could even decide to have auto-rebalancing only when adding or removing brokers but not for both.
```

Comment:
The manual is always possible, even when you have auto configured for everything. Of course, it doesn't make sense because as soon as you do a scaling, and auto is configured, the rebalancing will start for you without the need for the manual one.